src/vaults/SectorCrossVault.sol

```diff
@@ -0,0 +1,344 @@
+					uint16(msgType.REQUESTREDEEM),
+					uint16(block.chainid)
+				);
+			}
+
+			unchecked {
+				i++;
+			}
+		}
+	}
+
+	function redeemFromVaults(address[] calldata vaults, uint256[] calldata shares)
+		public
+		onlyRole(MANAGER)
+		checkInputSize(vaults.length, shares.length)
+	{
+		for (uint256 i = 0; i < vaults.length; ) {
+			Vault memory tmpVault = depositedVaults[vaults[i]];
+
+			if (tmpVault.allowed) revert VaultNotAllowed(vaults[i]);
+
+			if (tmpVault.adapter == address(0)) {
+				BatchedWithdraw(vaults[i]).redeem(shares[i], address(this), address(this));
+			} else {
+				IXAdapter(tmpVault.adapter).sendMessage(
+					shares[i],
+					vaults[i],
+					address(this),
+					tmpVault.chainId,
+					uint16(msgType.REDEEM),
+					uint16(block.chainid)
+				);
+			}
+			// Not sure if it should request manager intervention after redeem when in different chains
+
+			unchecked {
+				i++;
+			}
+		}
+	}
+
+	// Not sure if caller has to pass array of vaults
+	// Can be dangerous if manager fails or forgets an address
+	// TODO asks loaner
+	function harvestVaults(address[] calldata vaults) public onlyRole(MANAGER) {
+		uint256 depositValue = 0;
+
+		if (harvestLedger.isOpen) revert OnGoingHarvest();
+
+		for (uint256 i = 0; i < vaults.length; ) {
+			Vault memory tmpVault = depositedVaults[vaults[i]];
+
+			if (tmpVault.adapter == address(0)) {
+				depositValue +=
+					BatchedWithdraw(vaults[i]).balanceOf(address(this)) *
+					BatchedWithdraw(vaults[i]).withdrawSharePrice();
+			} else {
+				IXAdapter(tmpVault.adapter).sendMessage(
+					0,
+					vaults[i],
+					address(this),
+					tmpVault.chainId,
+					uint16(msgType.REQUESTVALUEOFSHARES),
+					uint16(block.chainid)
+				);
+
+				harvestLedger.request.push(Request(block.timestamp, tmpVault.chainId, vaults[i]));
+			}
+			unchecked {
+				i++;
+			}
+		}
+
+		harvestLedger.depositValue = depositValue;
+		harvestLedger.isOpen = true;
+	}
+
+	function finalizeHarvest(uint256 expectedValue, uint256 maxDelta) public onlyRole(MANAGER) {
+		HarvestLedger memory hLedger = harvestLedger;
+
+		// Compute actual tvl
+		uint256 actualValue = 0;
+
+		if (!hLedger.isOpen) revert HarvestNotOpen();
+
+		// Get all values from message board
+		uint256 i = hLedger.openIndex;
+		while (i < hLedger.request.length) {
+			Vault memory tmpVault = depositedVaults[hLedger.request[i].vault];
+
+			// If timestamp > message.timestamp transaction will revert
+			uint256 value = IXAdapter(tmpVault.adapter).readMessage(
+				hLedger.request[i].vault,
+				tmpVault.chainId,
+				hLedger.request[i].timestamp
+			);
+			actualValue += value;
+
+			unchecked {
+				i++;
+			}
+		}
+
+		// Check if tvl is expected before commiting
+		uint256 delta = expectedValue > actualValue
+			? expectedValue - actualValue
+			: actualValue - expectedValue;
+		if (delta > maxDelta) revert SlippageExceeded();
+
+		// Commit values
+		withdrawSharePrice = (hLedger.depositValue + actualValue) / totalSupply();
```

Comment:
this should be done via `_processWithdraw` so that the timestamp is updated also

Reply:
Done.